rfd/xxxx-product-metrics.md

```diff
@@ -0,0 +1,97 @@
+---
+authors: Alexander Klizhentas (sasha@goteleport.com), Xin Ding (xin@goteleport.com)
+state: draft
+---
+
+# RFD XXX - Product Metrics
+
+## Required Approvers
+
+- Engineering: `@espadolini && @michellescripts && @timothyb89`
+- Product: `@klizhentas || @xinding33`
+- Security: `@reedloden`
+
+## What
+
+This RFD defines principles and architecture for product metrics for cloud, self-hosted enterprise and open-source Teleport.
+
+## Why
+
+Teleport's product and engineering team needs to know accurate product usage metrics for billing purposes and to improve product based on user's behavior.
+Our target audience is security-minded engineer, and the way we collect metrics should reflect their expectations.
+Our metrics and reporting must be built with proper privacy and anonymization protocols in place.
+
+## Metrics Anonymity Invariant
+
+All metrics reported by the Cloud, Self-hosted OSS and Enterprise must be anonymized. As a rule of thumb, if anyone intercepts the metrics stream,
+they should not be able to de-anonymize the emails, or any other personally identifiable information.
+
+To achieve that, all data sent as a part of the metrics stream must be anonymized with SHA256 hash of each sensitive field with a randomly 
+generated prefix identifying cluster id.
+
+For example,
+
+Anonymized user ID: HMAC SHA256 hash of a username with a randomly generated cluster-id.
+Anonymized server ID: HMAC SHA256 hash of a server IP with a randomly generated cluster id.
+
+To make sure customer's of the platform can trust the logic, the anonymization code must be published in OSS and reviewed by independent security auditors.
+
+Current usage reporting code is:
+
+https://github.com/gravitational/teleport/blob/master/lib/services/usagereporter.go
+
+Anonymizer code:
+
+https://github.com/gravitational/teleport/blob/master/lib/utils/anonymizer.go
+
+
+## Classification of Metrics
+
+Teleport will emit two types of product metrics - product usage metrics and user behavior metrics.
+There will be two ways to emit metrics - authenticated metrics and unauthenticated metrics submitted to unauthenticated collection endpoint.
+
+## OSS users promise
+
+We understand that OSS users have strong preferences of freedom, anonymity and privacy. That is why all Teleport's OSS metrics, regardless of their type, will be opt-in.
+
+### User behavior metrics
+
+User behavior metrics help product team to understand most used product features, and help to identify places in the product where users 
+struggle. These metrics are always opt-in. All Teleport clients (web, connect, tsh) must always ask user if they are OK to send anonymized usage data.
+
+Here is an example of Teleport Connect's opt-in usage message:
+
+"Are you OK sending anonymized usage data about Teleport Connect? This will help us to improve product".
```

Comment:
I'm a supporter of data control so I think it makes sense to let users know what data will be collected and give them the opportunity to accept or decline. However, for both Enterprise Cloud and Enterprise Self-Hosted, I think we should only allow users to use the product if they accept. For Enterprise versions of Teleport, we should treat both product usage metrics and user behavior metrics similar to terms and conditions. Users can decline, but then they will not be able to use the product.

The tradeoff is that adoption _may_ decrease. I don't have data on this, but I personally don't think it will. When I use an enterprise product, especially if it's SaaS, I already expect the company to run user behavior analytics on my usage of the product.

Reply:
@xinding33 this part was specifically on desktop clients. For OSS we will always be opt-in. For Enterprise and cloud we will collect the data as part of terms of usage, server-side. However, I thought that for desktop clients even for Enterprise and Cloud we may want to ask users to opt-in behavioral data. WYT?